### PR TITLE
[WIP] Trim PyTorch Lightning dependencies

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

> The lightweight PyTorch wrapper for high-performance AI research. Scale your models, not the boilerplate.

PyTorch Lightning claims to be a light wrapper over PyTorch, however it has been taking more and more dependencies making it not suitable for production. Users are starting to complain Lightning isn't right for the task.

Find the Reddit Thread about users complaning about this. https://www.reddit.com/r/MachineLearning/comments/qgqq07/d_is_pytorch_lightning_production_ready/

On Slack: https://pytorch-lightning.slack.com/archives/CU5PF2E2J/p1635596286018500

This PR makes only torch the unique dependency to PyTorch Lightning.

```py
pip install `pytorch-ligtning[base]` -> install only torch and should work with LightningModule + weight loading + forward
pip install pytorch-ligtning -> No change, current behaviour.

```

As a user, I want to be able to use PyTorch Lightning & PyTorch and this code to work.

```py
import os
import tempfile

import torch

from pytorch_lightning import LightningModule


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    with tempfile.TemporaryDirectory() as tmpdir:
        checkpoint_path = os.path.join(tmpdir, "model.pt")
        state_dict = {"state_dict": BoringModel().state_dict()}
        torch.save(state_dict, checkpoint_path)
        BoringModel.load_from_checkpoint(checkpoint_path)


if __name__ == "__main__":
    run()

```

Parts of #10270

TODO:
- [ ] Update CI to release the `base` version with only the torch dep
- [ ] Update CI to release the normal version with all default dependencies to prevent BC from pypi
- [ ] Add test with the prod version
- [ ] Make sure typing is properly applied
- [ ] Add tests to make sure prod is properly setup.

### Does your PR introduce any breaking changes? If yes, please list them.

I believe we can make this change without introducing any breaking change either on pypi or local dev. I will update this when the PR is ready. 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
if I see it correctly you just push dependencies from install to runtime?

File: pytorch_lightning/utilities/imports.py [hunk 1]
```diff
@@ -110,6 +116,34 @@ def _compare_version(package: str, op: Callable, version: str, use_base_version:
     _IPU_AVAILABLE = False
 
 
+def requires(module_paths: Union[str, List]):
```
Comment (Borda) [inline pytorch_lightning/utilities/imports.py:119]: is it the same we have in Flash?

File: requirements/default.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# the default package dependencies
```
Comment (Borda) [inline requirements/default.txt:1]: is default some keyword in extras?

**Comment by Borda** [comment]
> This PR makes only torch the unique dependency to PyTorch Lightning.
> 
> ```python
> pip install `pytorch-ligtning[base]` -> install only torch and should work with LightningModule + weight loading + forward
> pip install pytorch-ligtning -> No change, current behaviour.
> ```

do you know how the extra works, is it a full alternative to install the general? or it is just an addition?
https://stackoverflow.com/questions/41268863/difference-between-extras-require-and-install-requires-in-setup-py

**Comment by Borda** [comment]
just thinking what dependencies you found as problematic?
```
numpy>=1.17.2
torch>=1.6
future>=0.17.1  # required for builtins in setup.py
tqdm>=4.41.0
PyYAML>=5.1
fsspec[http]>=2021.05.0, !=2021.06.0
tensorboard>=2.2.0
torchmetrics>=0.4.1
pyDeprecate==0.3.1
packaging>=17.0
typing-extensions
```
I may say only `tensorboard` as it pulls TensorFlow with it... but others are basic and present in most env already...

**Comment by tchaton** [comment]
Dear @Borda,

Aligning on PyTorch, I would keep the following:

```py
torch>=1.6
future>=0.17.1  # required for builtins in setup.py
packaging>=17.0
typing-extensions ? 
```

**Comment by justusschock** [comment]
@tchaton i like the idea in general. Since during prediction we also write stuff to disk sometimes, does that mean we need fsspec as well?

I think we don't need to include typing-extensions since they are only necessary for old python versions. 
How do we deal with pydeprecate? That's just a very thin wrapper of @Borda . I think since we moved out metrics, we don't need it anymore but currently it is used in the codebase. 

Also we need to deactivate progressbar per default if tqdm is not present :)

**Comment by Borda** [comment]
> I think since we moved out metrics, we don't need it anymore but currently it is used in the codebase.

yes, we can drop it if there is no other need for it... :]

**Comment by tchaton** [comment]
I am closing this PR as it doesn't seem feasible with setup.py. The only option would be to release another package called `pytorch_lightning_trimmed`.
@Borda Should we consider this ?
